proposals/NNNN-swift-distributed-tracing.md

```diff
@@ -0,0 +1,306 @@
+
+## Detailed design
+
+Both library/framework authors and end-users use the same API to create tracing spans. Adding a span to functions can
+be achieved by using the `withSpan(operationName:context:body:)` API:
+
+```swift
+func performOperation() -> String {
+    withSpan("operation") {
+        return "handled"
+    }
+}
+```
+
+### Customizing the span
+
+A span can be further customized using `withSpan(operationName:context:body:)` as well:
+
+```swift
+func handleRequest(_ request: Request) -> Response {
+    withSpan(request.endpointPath, ofKind: .server) { span in
+        span.attributes["http.method"] = request.method.rawValue
+        return respond(to: request)
+    }
+}
+```
+
+### Recording errors
+
+Throwing can be handled by either recording errors manually into a span by calling `span.recordError(error:)`,
+or by wrapping a potentially throwing operation using the `withSpan(operationName:context:body:)` function,
+which automatically records any thrown error.
+
+```swift
+func handleRequest(_ request: Request) throws -> Response {
+    try withSpan(request.endpointPath, ofKind: .server) {
+        try respond(to: request)
+    }
+}
+```
+
+### Swift Concurrency
+
+Swift Distributed Tracing ships both synchronous and asynchronous versions of `withSpan(operationName:context:body:)`.
+In both cases, spans are automatically ended at the end of the body closure scope:
+
+```swift
+func handleRequest(_ request: Request) async throws -> Response {
+    try await withSpan(request.endpointPath, ofKind: .server) {
+        try await respond(to: request)
+    }
+}
+```
+
+### Child Spans
+
+A distributed tracing span may have one or more child spans. While Swift Distributed Tracing also allows for manually
+linking spans, creating a child span is easiest when using the
+task-local [`ServiceContext`](https://github.com/apple/swift-service-context) behind the scenes:
+
+```swift
+func handleRequest(_ request: Request) async throws -> Response {
+    try await withSpan(request.endpointPath, ofKind: .server) {
+        let user = try await fetchUser(byID: request.userID)
+        return try await respond(to: request)
+    }
+}
+
+func fetchUser(byID id: String) async throws -> User {
+    try await withSpan("fetchUser") { span in
+        span.attributes["user.id"] = id
+    }
+}
+```
+
+In this example, the `withSpan(operationName:context:body:)` call within `handleRequest(_:)` creates a span.
+When creating the span, the tracer implementation sets trace and span identifiers on the task-local
+[ServiceContext](https://github.com/apple/swift-service-context). `fetchUser(id:)` will then automatically "inherit"
+these identifiers, indicating to the tracer implementation that the "fetchUser" span should be a child of the request
+span.
+
+### Low-level API
+
+When using Swift NIO `EventLoopFuture`s or similar patterns, we need to control the spans lifetime manually. This can be
```

Comment:
Maybe mention the service context will need to be propagated manually if using EventLoopFutures.

Reply:
```suggestion
When using Swift NIO `EventLoopFuture`s or similar patterns, we need to control the spans lifetime manually. 
The service context will also need to be propagated manually by passing it to the context parameter of startSpan, 
because event loop futures cannot automatically propagate task local values.
This can be
```